Make LoadBalancer creation more self-healing; don't delete on AWS

justinsb opened the pull request:
We can't delete the load-balancer & recreate it all the time; we risk losing the
endpoint (on AWS) and/or service disruption.

The GCE provider looks to be mostly idempotent.  But we need to start by requiring
that it is idempotent.  The OpenStack provider probably needs a once-over.

justinsb commented:
It looks like CreateExternalLoadBalancer is supposed to be idempotent, given the GCE implementation.  

Also, I think we need this to be idempotent, because we otherwise have no other repair mechanism.

I do not think that it is entirely idempotent, but I think we have to try to get there, and we should write code assuming that the providers will get there.

brendandburns commented on pkg/cloudprovider/gce/gce.go:
```diff
@@ -309,6 +309,7 @@ func translateAffinityType(affinityType api.AffinityType) GCEAffinityType {
 // CreateTCPLoadBalancer is an implementation of TCPLoadBalancer.CreateTCPLoadBalancer.
 // TODO(a-robinson): Don't just ignore specified IP addresses. Check if they're
 // owned by the project and available to be used, and use them if they are.
+// TODO: Verify this is idempotent
```
afaik, this is not idempotent, it will fail (innoucously) with an "already exists" error, but it should definitely throw an error.

justinsb replied:
Oh - I thought that was what each check to ignore StatusConflict was doing?

I'll add a item to my todo list for an e2e test that requires changing a loadbalancer (I think adding another port ought to do it)

a-robinson replied:
No, Justin was right, this is idempotent in that it assumes an already existing resource is a success - see the handling of http.StatusConflict errors.

You can get rid of the TODO though, unless your intent is for us to add automated verification.

a-robinson replied:
Sorry, I very well may be missing something about it - what do you see in it that might not be idempotent?

brendandburns commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -225,44 +225,25 @@ func (s *ServiceController) createLoadBalancerIfNeeded(namespacedName types.Name
 				return err, retryable
 			}
+		} else {
+			glog.Infof("Not creating LB for service %s that doesn't want one.", namespacedName)
```
This log seems spammy V(4) it?

bgrant0607 commented:
cc @a-robinson

a-robinson commented on pkg/cloudprovider/openstack/openstack.go:
```diff
@@ -480,6 +480,7 @@ func (lb *LoadBalancer) GetTCPLoadBalancer(name, region string) (endpoint string
 // a list of regions (from config) and query/create loadbalancers in
 // each region.
 
+// TODO: Verify this is idempotent
```
There's relevant discussion ongoing around this:
https://github.com/GoogleCloudPlatform/kubernetes/pull/7852#discussion_r29817700
https://github.com/GoogleCloudPlatform/kubernetes/issues/8352

a-robinson commented on pkg/cloudprovider/cloud.go:
```diff
@@ -62,7 +62,7 @@ type TCPLoadBalancer interface {
 	// if so, what its IP address or hostname is.
 	GetTCPLoadBalancer(name, region string) (endpoint string, exists bool, err error)
-	// CreateTCPLoadBalancer creates a new tcp load balancer. Returns the IP address or hostname of the balancer
+	// CreateTCPLoadBalancer idempotently creates or updates a tcp load balancer. Returns the IP address or hostname of the balancer
```
Having Create handle updating as well seems like a high bar - one that certainly won't work with the current GCE implementation. Is there a reason that it's needed vs Create only being responsible for the initial creation and relying on the service controller to handle the necessary update logic?

a-robinson commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -225,44 +225,25 @@ func (s *ServiceController) createLoadBalancerIfNeeded(namespacedName types.Name
 		glog.Infof("LB already exists and doesn't need update for service %s", namespacedName)
 		return nil, notRetryable
 	}
-	if cachedService != nil {
```
Getting rid of this deletion here will surely break GCE at head, because GCE's CreateTCPLoadBalancer function doesn't currently handle updating the LB's parameters.

a-robinson replied:
To be more specific - it'll break us in the case where some aspect of the service definition has changed (such as the port numbers) that requires modifying the load balancer, which we currently implement by tearing it down and recreating it.

a-robinson commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -225,44 +225,25 @@ func (s *ServiceController) createLoadBalancerIfNeeded(namespacedName types.Name
 		if err != nil {
 			return fmt.Errorf("Error getting LB for service %s", namespacedName), retryable
 		}
-		if exists && stringSlicesEqual(service.Spec.PublicIPs, []string{endpoint}) {
```
Getting rid of the endpoint comparison for non-cached clusters will also reduce correctness in cases where we don't have a service cached and the existing LB for the service is out of sync with what k8s knows about the service.

a-robinson commented on pkg/cloudprovider/servicecontroller/servicecontroller.go:
```diff
@@ -225,44 +225,25 @@ func (s *ServiceController) createLoadBalancerIfNeeded(namespacedName types.Name
+			glog.Infof("Deleting LB for service %s", namespacedName)
 			if err := s.ensureLBDeleted(service); err != nil {
 				return err, retryable
 			}
```
I don't think you want to fall through here if the deletion is successful - you'll go on to create a LB for the service even though it doesn't want one. You basically want to get rid of the else on the line below this.

a-robinson commented:
I want to make sure I understand the motivation behind the service controller changes - is your intent that every cloud provider implements its own version of the logic that's currently in the controller for making sure that what currently exists is correct and that we update it when needed?

justinsb commented:
So the motivation is that I think the cloud providers can often do better than a delete and create when making changes to a load balancer.  And on everyone's favorite cloud (AWS), when you delete a LB and recreate it, it has a different endpoint (i.e. the user must reconfigure their DNS).  I suspect that other clouds may not be quite as bad, but they will likely have service disruption during the time in between the delete and the create.

So I think it makes sense to put this logic in to the cloud provider; it can always just implement a delete & recreate.  But on AWS at least, I'll have to change the existing LB.

Also, making sure this is idempotent makes retries safer; I do think the GCE code looks idempotent.  If we put that in the function spec, then we can be a little more naive in servicecontroller.go by pushing the logic into the cloud provider, and these are the changes I came up with.

I propose we:

* Decide whether the LB implementation should be able to change existing LBs (even if it implements this as a delete & create).  I think yes, because AWS needs it, and I think it is better on GCE.  I don't know how much we need it now (adding ports to an existing service seems pretty niche), but I can see this becoming more common as we make the model richer.
* Decide whether we want to require the LB create to be idempotent (I think so, as otherwise we're likely to end up with "stuck" LBs in case of failure, and I think GCE already does this)
* Then I can look again at servicecontroller.go, to see if I can work around the issues you identified, under the assumption that the LB create will idempotently create or update the named LB to the state requested.  In particular, I don't want an update to involve a delete & create unless there is no alternative.

a-robinson commented:
Sounds good, thanks for making sure we're on the same page.

We're coming in pretty hot toward 1.0, so while I completely agree with the direction, we'd have to move fast to get the required changes in. I'm happy to help, but it's unlikely to be my top priority over the next couple weeks. However, as long as the API is where we want it, this could pretty easily be improved post-1.0.

> Decide whether the LB implementation should be able to change existing LBs (even if it implements this as a delete & create). I think yes, because AWS needs it, and I think it is better on GCE. I don't know how much we need it now (adding ports to an existing service seems pretty niche), but I can see this becoming more common as we make the model richer.

Agreed on both counts -- that it allows for the optimal experience on each cloud, and also that it's a somewhat niche use case at the moment.

> Decide whether we want to require the LB create to be idempotent (I think so, as otherwise we're likely to end up with "stuck" LBs in case of failure, and I think GCE already does this)

Agreed

> Then I can look again at servicecontroller.go, to see if I can work around the issues you identified, under the assumption that the LB create will idempotently create or update the named LB to the state requested. In particular, I don't want an update to involve a delete & create unless there is no alternative.

Agreed - this should come after the cloud providers have added a function that supports what we need.

justinsb commented:
Changing status to WIP; there are lots of great comments here and I'm going to have another crack at this.  I think this will be the last problem in the e2e test suite for AWS (that the AWS LB endpoint changes when you change e.g. a port, because the only mechanism we have is to delete & recreate the LB)

brendandburns commented:
@justinsb let me know when this is ready for review again.

Thanks!
--brendan

justinsb committed:
Add more events to LoadBalancer management

It would be nice to see these events in the service like we do with pods;
also we've had trouble here so a few more events would be handy!

justinsb committed:
Change CreateTCPLoadBalancer -> EnsureTCPLoadBalancer; implementations auto-delete if already exists

Previously the servicecontroller would do the delete, but by having the cloudprovider
take that task on, we can later remove it from the servicecontroller, and the
cloudprovider can do something more efficient.

justinsb committed:
servicecontroller relies on cloudprovider to delete LB if needed

We previously made the cloudproviders take on the responsibility of deleting
existing load balancers; this lets us simplify the servicecontroller logic
and also lays the groundwork for more efficient cloudprovider LB implementations
to do an in-place change on a LB.

justinsb committed:
AWS: Make load balancer creation idempotent on AWS

This turned out to be a little convoluted, but is needed because deleting an ELB on AWS
is a painful UX - it won't have the same endpoint when it is recreated.

Also started splitting the provider into files, but only for new functions (so far!)

justinsb commented:
I think - finally - this is ready for re-review!

I believe the logic in the servicecontroller to be more "self-evidently correct" now (although now I'm sure there will be a horrendous & obvious error I've missed).  The servicecontroller could be _much_ simpler, except that we want sensible events.

Essentially, we push down the delete from the servicecontroller into the cloudprovider, and simplify the servicecontroller accordingly.

Then, because AWS ELB will change DNS name every time you delete the ELB, we update AWS ELBs in place.  It would be pretty easy to do this for GCE as well - I think it is pretty close actually.   The AWS ELB logic is pretty tedious, but there's no way around that.

thockin commented on pkg/cloudprovider/gce/gce.go:
```diff
@@ -335,8 +335,23 @@ func translateAffinityType(affinityType api.ServiceAffinity) GCEAffinityType {
 // CreateTCPLoadBalancer is an implementation of TCPLoadBalancer.CreateTCPLoadBalancer.
```
comments are stale

thockin commented on pkg/util/set.go:
```diff
@@ -129,6 +129,21 @@ func (s1 StringSet) IsSuperset(s2 StringSet) bool {
+// Equal returns true iff s1 is equal (as a set) to s2.
+// Two sets are equal if their membership is identical.
+// (In practice, this means same elements, order doesn't matter)
+func (s1 StringSet) Equal(s2 StringSet) bool {
```
test?

thockin commented:
Is this really critical pre-1.0?

justinsb commented:
Probably not critical pre-1.0.  On AWS, the behavior when you change settings on a LoadBalancer service is a little obnoxious, in that it deletes your ELB and then creates a new one with a new CNAME.  But that is probably not terrible common anyway.

I was hoping this would fix some of the weirdness around LB creation that we were seeing, but I didn't find any obvious bugs.  I found one really small and mostly theoretical edge-case, and we fixed that.

I'm happy to tag this as post-v1; if I get a log from a kube-controller-manager and it shows that there is a problem in this code we can maybe consider this approach.

thockin commented:
SGTM thanks

On Thu, Jun 18, 2015 at 3:34 PM, Justin Santa Barbara <
notifications@github.com> wrote:

> Probably not critical pre-1.0. On AWS, the behavior when you change
> settings on a LoadBalancer service is a little obnoxious, in that it
> deletes your ELB and then creates a new one with a new CNAME. But that is
> probably not terrible common anyway.
>
> I was hoping this would fix some of the weirdness around LB creation that
> we were seeing, but I didn't find any obvious bugs. I found one really
> small and mostly theoretical edge-case, and we fixed that.
>
> I'm happy to tag this as post-v1; if I get a log from a
> kube-controller-manager and it shows that there is a problem in this code
> we can maybe consider this approach.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/8366#issuecomment-113306238>
> .
>

brendandburns commented:
@justinsb is this still alive?

justinsb commented:
It is still alive; I think it is pretty important - see #11002 which I opened to describe the exact issue.  But I think I need to take another look at it now that we are post v1 (at the very least to clean up merge conflicts).

justinsb committed:
Add unit test for StringSet Equal

justinsb committed:
e2e: Test load-balanced service endpoint preserved

Previously we weren't preserving the service endpoint on a load-balanced
service, at least on AWS.  The test had to test reality, not
aspirations!

Now we should have fixed this, so we can revert the e2e test to check
that the service endpoint is indeed preserved.

Fixes #11002

justinsb commented:
Rebased & fixed up some errors & incorporated feedback.  Also added e2e test that should verify that this is all working as intended.

Running a full e2e test on AWS now.

justinsb commented:
I think that is a flake (it is a service test, but it is a Type=Cluster service).

```
• Failure [86.040 seconds]
Services
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:947
  should serve multiport endpoints from pods [It]
  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:231

  Timed out waiting for service multi-endpoint-test in namespace e2e-tests-service-0-uoqlh to expose endpoints map[podname1:[100]] (1m0s elapsed)

  /go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:1096
```

justinsb commented:
@k8s-bot test this please

justinsb commented:
Ironically this test is failing on AWS (I guess that is a good thing; it means the test is real!).  Investigating.

justinsb committed:
AWS: Configure LoadBalancer health checks

ELB will automatically create a health check, but if we update the
listeners the old health check port sticks around, and all the instances
are marked offline.

Update the health-checks to match the listeners: we just check the first
valid service port, with some hard-coded options for timeouts / retries etc.

justinsb commented:
So this is now working on AWS.  The problem was that ELB automatically (and undocumentedly AFAICT) creates a health-check rule matching your container port.  But if you change your container port, it then does not update the rule.  So ELB was marking all the instances as down.

Added code to sync the health checks.

Finally ready for review for merging!

benmccann commented:
@justinsb looks like this needs a rebase

justinsb committed:
Update imports to new k8s.io style

justinsb committed:
Move new files into pkg/cloudprovider/providers/aws/

They were not caught by the general reorg because they weren't on
master.
